README
ssh-sc-auth - SSH SmartCard Authentication

ssh-sc-auth is a smart-card authentication handler for use with the
AuthorizedKeysCommand feature of SSH.  The program will return a dynamic
authorized_keys entries for the user, based on either sshPublicKey or
SSL certificates stored in a remote LDAP server.


INSTALLATION
------------

# Debian:
apt-get -y install              \
    libconfig-general-perl      \
    libcrypt-openssl-rsa-perl   \
    libcrypt-openssl-x509-perl  \
    libnet-ldap-perl            \
    libproc-processtable-perl   \
    libtest-perl-critic-perl

# CentOS:
yum -y install                  \
    perl-Module-Build           \
    perl-Archive-Tar            \
    perl-Config-General         \
    perl-Crypt-OpenSSL-RSA      \
    perl-Crypt-OpenSSL-X509     \
    perl-LDAP                   \
    perl-Proc-ProcessTable      \
    perl-Test-Perl-Critic

perl Build.PL
./Build
./Build test
./Build install

Note: you should create a user who will run ssh-sc and
ensure that /opt/ssh-sc files are only owned by root,
group readable by the AuthorizedKeyCommandUser.

Example:
    useradd -d /opt/ssh-sc -s /usr/sbin/nologin ssh-auth
    chown -R root.ssh-auth /opt/ssh-sc
    find /opt/ssh-sc -type d -exec chmod 750 {} \;
    find /opt/ssh-sc -type f -exec chmod 640 {} \;
    chmod 750 /opt/ssh-sc/sbin/*


UNINSTALL
---------

./Build uninstall



OPENSSH VERSION
---------------

You will need a version of OpenSSH that provides the AuthorizedKeysCommand
feature.  I believe this is anything >= v6.6.1.  Debian Jessie and CentOS 7
are known to have a useable version of OpenSSH.


SSHD_CONFIG
-----------

    PubkeyAuthentication yes
    AuthorizedKeysFile /etc/ssh/ssh_authorized_keys
    AuthorizedKeyscommand /opt/ssh-sc/sbin/ssh-sc-auth
    AuthorizedKeysCommandUser ssh-auth
    AgentForwarding no
    MaxSessions 1


AUTHORIZED_KEYS FILES
---------------------

sshd will still load keys from the AuthorizedKeysFile even though
AuthorizedKeysCommand has been specified.  It is important to set
AuthorizedKeysFile to an approved, non-modifiable file in order to
prevent users from using their own keys (presumably passwordless).

    touch /etc/ssh/ssh_authorized_keys

    # sshd_config
    ...
    AuthorizedKeysFile /etc/ssh/ssh_authorized_keys
    ...

This will still allow the use of approved keys that are not part of
the organization's SmartCard framework.


SSH-AGENT AND AGENT-FORWARDING
------------------------------

The use of ssh-agent allows one to cache credentials in order to save the user
from having to constantly enter passwords.  While it is unclear whether it is
ssh-agent caching the user's pin or if it is OpenSC keeping the card unlocked, the
same functionality exists when a smart-card is in use.  In a nutshell, a
compromised client running ssh-agent will allow the attacker to gain SSH
access to remote systems without requiring the knowledge of the PIN or even
having possession of the smart-card.  This gets further exacerbated when agent
forwarding is in use as smart-card authentication is chained back to the
user's system, which wouldn't require the PIN per the reasons outlined above.

There is no assurance that the connection is from a user with physical
possession of the card and knowledge of the PIN or is from an attacker
leveraging the victim's use of ssh-agent or agent-forwarding.


   RECOMMENDATION
   ==============

   * Run ssh-sc-auth only on a bastion host and require password and/or
     some OTP authentication, such as SecurID, at all internal hosts.

   * Disable agent forwarding by adding no-agent-forwarding to the
     'ssh-options' configuration option of ssh-sc-auth and setting
     AllowAgentForwarding to 'no' in sshd_config.

   * Run a restrictive or chrooted shell on the bastion so that the
     user does not actually have command-line access to the bastion.


CONTROL MASTER
--------------

The use of ControlMaster enables the sharing of multiple sessions over a
single connection.  When in use, the first connection requires authentication
and subsequent connections do not.  A compromised client using ControlMaster
will allow the attacker to gain SSH access to the remote systems without
requiring the knowledge of the PIN or even having possession of the
smart-card.

There is no assurance that the connection is from a user with physical
possession of the card and knowledge of the PIN or is from an attacker
leveraging the victim's use of ControlMaster.

   
   RECOMMENDATION
   ==============

   * Set MaxSessions to 1 in sshd_config.



debian/changelog
ssh-sc-auth (1.0.0-1) unstable; urgency=low

  * Initial Release.

 -- Bennett Samowich <bennett@foolean.org>  Tue, 17 Oct 2017 06:18:05 -0400


Build.PL
use strict;
use warnings;

use Data::Dumper;
use Digest::SHA;
use File::Copy;
use File::Path qw( make_path rmtree );
use Module::Build;

our $VERSION = '1.0.0';

our $INSTALL_BASE = '/opt/ssh-sc';

my $class = Module::Build->subclass( code => <<'EOF');
    sub make_dir {
        use File::Path qw( make_path );
        
        my ( $dirs, $mode ) = @_;

        for my $dir ( @$dirs ) {
            if ( ! -d $dir ) {
                printf "make_path %s\n", $dir;
                make_path $dir;
            }
            chmod $mode, $dir;
        }
    }

    sub ACTION_install {
        my $self = shift;

        # Account for destdir
        $self->install_base($self->destdir() . $self->install_base());

        # Create the directory structure
        make_dir (
            [
                $self->install_base,
                $self->install_base . '/sbin',
                $self->install_base . '/doc',
                $self->install_base . '/doc/selinux',
                $self->install_base . '/etc',
                $self->install_base . '/etc/ssl',
                $self->install_base . '/man',
                $self->install_base . '/man/man1',
                $self->install_base . '/man/man5',
            ], 0750,
        );

        # Call the parent install function
        $self->SUPER::ACTION_install;
    }

    sub ACTION_uninstall {
        my $self = shift;

        # Account for destdir
        $self->install_base($self->destdir() . $self->install_base());

        if ( -d $self->install_base() ) {
            printf "Removing %s\n", $self->install_base();
            File::Path::rmtree($self->install_base(), 1, 1);
        }
    }

    sub process_sbin_files {
        my $self = shift;

        if ( $self->invoked_action() eq 'install' ) {

            my $files = $self->{'properties'}->{'sbin_files'};
            while ( my ( $src, $dst ) = each %$files ) {
                printf "Installing %s/%s\n", $self->install_base(), $dst;
                File::Copy::copy( $self->base_dir() . '/' . $src, $self->install_base() . '/' . $dst);
                chmod 0750, $self->install_base() . '/' . $dst;
            }

            # Search for POD and create man1 pages
            use Pod::Man;
            use Pod::Find qw(pod_find);
            use File::Basename;
            my $man1dir      = $self->install_base() . '/man/man1';
            my $dist_version = $self->dist_version();
            my %pods = pod_find({},( $self->base_dir() . '/sbin'));
            foreach my $pod ( keys %pods ) {
                my $pod_rel = $pod;
                my $base_dir = $self->base_dir();
                $pod_rel =~ s/${base_dir}\///mx; 
                my $man1page = $man1dir . '/' . basename($pod, '.pod') . '.1';
                printf "Manifying %s -> %s\n", $pod_rel, $man1page;
                my $parser = Pod::Man->new(
                    release => "ssh-sc v$dist_version", section => 1
                );
                $parser->parse_from_file( $pod, $man1page );
                chmod 0640, $man1page;
            }
        }

    }

    sub process_etc_files {
        my $self = shift;

        use Digest::SHA;

        if ( $self->invoked_action() eq 'install' ) {
            my $files = $self->{'properties'}->{'etc_files'};
            while ( my ( $src, $dst ) =  each %$files ) {

                if ( -f $self->install_base() . '/' . $dst ) {
                    printf "Skipping %s/%s (pre-existing)\n", $self->install_base(), $dst;
                    next;
                }

                printf "Installing %s/%s\n", $self->install_base(), $dst;
                File::Copy::copy( $self->base_dir() . '/' . $src, $self->install_base() . '/' . $dst);
                chmod 0640, $self->install_base() . '/' . $dst;
            }
        
            # Search for POD and create man5 pages
            use Pod::Man;
            use Pod::Find qw(pod_find);
            use File::Basename;
            my $man5dir      = $self->install_base() . '/man/man5';
            my $dist_version = $self->dist_version();
            my %pods = pod_find({},( $self->base_dir() . '/etc'));
            foreach my $pod ( keys %pods ) {
                my $pod_rel = $pod;
                my $base_dir = $self->base_dir();
                $pod_rel =~ s/${base_dir}\///mx; 
                my $man5page = $man5dir . '/' . basename($pod, '.pod') . '.5';
                printf "Manifying %s -> %s\n", $pod_rel, $man5page;
                my $parser = Pod::Man->new(
                    release => "ssh-sc v$dist_version", section => 5
                );
                $parser->parse_from_file( $pod, $man5page );
                chmod 0640, $man5page;
            }
        }
    }
EOF

my $build = $class->new(
    dist_name         => 'ssh-sc',
    module_name       => 'ssh-sc',
    dist_version      => $VERSION,
    dist_author       => 'Bennett Samowich <bennett@foolean.org',
    dist_abstract     => 'SSH AuthorizedKeysCommand for SmartCard Authentication',
    extra_manify_args => { release => "ssh-sc v$VERSION" },
    license           => 'perl',
    install_base      => $INSTALL_BASE,
    sbin_files         => {
        'sbin/ssh-sc-auth' => 'sbin/ssh-sc-auth',
    },
    doc_files => {
        'README'                 => 'doc/README.dist',
        'selinux/ssh-sc-auth.te' => 'doc/selinux/ssh-sc-auth.te',
    },
    etc_files => {
        'etc/ssh-sc-auth.conf' => 'etc/ssh-sc-auth.conf',
    },
    requires  => {
        'Config::General'       => 0,
        'Crypt::OpenSSL::RSA'   => 0,
        'Crypt::OpenSSL::X509'  => 0,
        'File::Basename'        => 0,
        'Getopt::Long'          => 0,
        'IPC::Open3'            => 0,
        'Net::LDAPS'            => 0,
        'Proc::ProcessTable'    => 0,
        'Scalar::Util'          => 0,
        'Sys::Hostname'         => 0,
        'Sys::Syslog'           => 0,
        'strict'                => 0,
        'warnings'              => 0,
    },
    build_requires => {
        'Archive::Tar'       => 0,
        'Digest::SHA'        => 0,
        'Module::Build'      => 0,
        'Pod::Find'          => 0,
        'Pod::Man'           => 0,
        'Test::More'         => 0,
        'Test::Perl::Critic' => '>= 1.01',
    },
);

my $install_base = $build->install_base();
$build->install_path( 'sbin' => $install_base . '/sbin' );
$build->install_path( 'doc'  => $install_base . '/doc' );
$build->install_path( 'etc'  => $install_base . '/etc' );

$build->add_build_element('sbin');
$build->add_build_element('doc');
$build->add_build_element('etc');
$build->create_build_script;

# Announce what the user is to do next
printf "\nTo install '%s' 'v%s' into '%s', run:\n",
  $build->dist_name(), $build->dist_version, $build->install_base();
printf "\t./Build\n\t./Build test\n\t./Build install\n";
printf "\n";
printf "Note: you should create a user who will run %s and\n",
  $build->dist_name();
printf "ensure that %s files are only owned by root,\n",
  $build->install_base();
printf "group readable by the AuthenticationKeyCommandUser.\n\n";
printf "Example:\n";
printf "      useradd -d %s -s /usr/sbin/nologin ssh-auth\n",
  $build->install_base();
printf "      chown -R root.ssh-auth %s\n", $build->install_base();
printf "      find %s -type d -exec chmod 750 {} \\;\n",
  $build->install_base();
printf "      find %s -type f -exec chmod 640 {} \\;\n",
  $build->install_base();
printf "      chmod 750 %s/sbin/*\n", $build->install_base();
